File: market_auto_start.py
"""
CENTRALIZED MARKET AUTO-START SYSTEM (Python Version)
======================================================

Alternative to Node.js version. Use this if you prefer Python.

Features:
- Auto-starts backend at 9 AM on market days
- Monitors health every 10 minutes
- Auto-restarts if backend crashes
- Logs all activity for debugging

Usage:
  1. Install: pip install APScheduler requests python-dateutil
  2. Deploy: Copy to /var/www/mytradingSignal/market_auto_start.py
  3. Run with PM2: pm2 start market_auto_start.py --interpreter python3
  4. Save: pm2 save && pm2 startup
"""

import schedule
import time
import subprocess
import sys
import os
import logging
from datetime import datetime, timedelta
from pathlib import Path
import requests
import json

# ============= CONFIGURATION =============

PROJECT_PATH = "/var/www/mytradingSignal"
BACKEND_NAME = "backend"
BACKEND_CMD = f"cd {PROJECT_PATH} && pm2 start backend/main.py --name {BACKEND_NAME} --interpreter python3"
LOG_DIR = Path("/var/log/mytradingSignal")
LOG_FILE = LOG_DIR / "market-auto-start.log"
ERROR_LOG = LOG_DIR / "market-auto-start-error.log"

# Market timings (IST)
MARKET_OPEN_TIME = "09:00"  # 9:00 AM
MARKET_CLOSE_TIME = "15:30"  # 3:30 PM
PREOPEN_END_TIME = "09:15"   # Pre-open ends at 9:15 AM
LIVE_START_TIME = "09:15"    # Live trading starts at 9:15 AM

# NSE/BSE Holidays
HOLIDAYS_2026 = [
    "2026-01-26", "2026-03-25", "2026-04-02", "2026-05-01",
    "2026-07-17", "2026-08-15", "2026-09-02", "2026-10-02",
    "2026-10-25", "2026-11-01", "2026-11-11", "2026-12-25"
]

HOLIDAYS_2027 = [
    "2027-01-26", "2027-03-14", "2027-04-02", "2027-05-01",
    "2027-07-06", "2027-08-15", "2027-08-19", "2027-10-02",
    "2027-10-16", "2027-10-20", "2027-10-21", "2027-12-25"
]

ALL_HOLIDAYS = HOLIDAYS_2026 + HOLIDAYS_2027

# ============= LOGGER SETUP =============

def setup_logger():
    """Configure logging to both file and console"""
    LOG_DIR.mkdir(parents=True, exist_ok=True)
    
    logger = logging.getLogger("market_scheduler")
    logger.setLevel(logging.DEBUG)
    
    # File handler for main log
    file_handler = logging.FileHandler(LOG_FILE)
    file_handler.setLevel(logging.DEBUG)
    
    # File handler for errors
    error_handler = logging.FileHandler(ERROR_LOG)
    error_handler.setLevel(logging.ERROR)
    
    # Console handler
    console_handler = logging.StreamHandler(sys.stdout)
    console_handler.setLevel(logging.INFO)
    
    # Formatter
    formatter = logging.Formatter(
        '%(asctime)s - %(name)s - %(levelname)s - %(message)s',
        datefmt='%Y-%m-%d %H:%M:%S'
    )
    
    file_handler.setFormatter(formatter)
    error_handler.setFormatter(formatter)
    console_handler.setFormatter(formatter)
    
    logger.addHandler(file_handler)
    logger.addHandler(error_handler)
    logger.addHandler(console_handler)
    
    return logger

logger = setup_logger()

# ============= MARKET FUNCTIONS =============

def is_market_day():
    """
    Check if today is a trading day
    Trading days: Monday-Friday, excluding holidays
    """
    now = datetime.now()
    day_of_week = now.weekday()  # 0 = Monday, 6 = Sunday
    
    # Skip weekends (5 = Friday, 6 = Saturday, 0 = Sunday becomes 7)
    if day_of_week >= 5:  # Saturday or Sunday
        return False
    
    # Check if it's a holiday
    date_str = now.strftime("%Y-%m-%d")
    if date_str in ALL_HOLIDAYS:
        logger.info(f"📅 Today is a market holiday: {date_str}")
        return False
    
    return True

def get_ist_time():
    """Get current time in IST (UTC+5:30)"""
    return datetime.now() + timedelta(hours=5, minutes=30)

def is_within_market_hours():
    """Check if current time is within market trading hours"""
    ist_time = get_ist_time()
    current_time = ist_time.strftime("%H:%M")
    
    return MARKET_OPEN_TIME <= current_time <= MARKET_CLOSE_TIME

def is_preopen_period():
    """Check if we're in pre-open period (9:00 - 9:15)"""
    ist_time = get_ist_time()
    current_time = ist_time.strftime("%H:%M")
    
    return MARKET_OPEN_TIME <= current_time < PREOPEN_END_TIME

def is_live_trading_period():
    """Check if we're in live trading period (9:15 - 3:30)"""
    ist_time = get_ist_time()
    current_time = ist_time.strftime("%H:%M")
    
    return LIVE_START_TIME <= current_time <= MARKET_CLOSE_TIME

# ============= BACKEND MANAGEMENT =============

def is_backend_running():
    """Check if backend is running via PM2"""
    try:
        result = subprocess.run(
            ["pm2", "list", "--silent"],
            capture_output=True,
            text=True,
            timeout=5
        )
        return BACKEND_NAME in result.stdout and "online" in result.stdout
    except Exception as e:
        logger.error(f"Failed to check PM2 status: {e}")
        return False

def start_backend():
    """Start backend using PM2"""
    try:
        logger.info("🚀 Starting backend...")
        subprocess.run(
            BACKEND_CMD,
            shell=True,
            check=True,
            capture_output=True,
            timeout=30
        )
        logger.info("✅ Backend started successfully")
        return True
    except Exception as e:
        logger.error(f"Failed to start backend: {e}")
        return False

def restart_backend():
    """Restart backend using PM2"""
    try:
        logger.info("🔄 Restarting backend...")
        subprocess.run(
            f"pm2 restart {BACKEND_NAME}",
            shell=True,
            check=True,
            capture_output=True,
            timeout=30
        )
        logger.info("✅ Backend restarted successfully")
        return True
    except Exception as e:
        logger.error(f"Failed to restart backend: {e}")
        return False

def health_check_backend():
    """Check if backend is responding to HTTP requests"""
    try:
        response = requests.get("http://localhost:8000/health", timeout=3)
        return response.status_code == 200
    except Exception as e:
        logger.debug(f"Health check failed: {e}")
        return False

# ============= MAIN SCHEDULER LOGIC =============

def check_and_start_backend():
    """Main function: Check if should start backend"""
    ist_time = get_ist_time()
    time_str = ist_time.strftime("%Y-%m-%d %H:%M:%S IST")
    
    print(f"\n🕐 Market Scheduler Check - {time_str}")
    
    # Check if market is open
    if not is_market_day():
        logger.debug("📊 Market closed (weekend or holiday)")
        return
    
    if not is_within_market_hours():
        logger.debug("⏰ Outside market hours")
        return
    
    # Log status
    logger.info("=" * 60)
    logger.info("=== MARKET DAY & TRADING HOURS ===")
    logger.info(f"🕐 Current Time (IST): {time_str}")
    
    if is_preopen_period():
        logger.info("📻 PRE-OPEN PERIOD (9:00-9:15): Values frozen, no live feed")
    else:
        logger.info("📈 LIVE TRADING PERIOD (9:15-3:30): Live values streaming")
    
    # Check backend status
    backend_running = is_backend_running()
    
    if not backend_running:
        logger.warning("🔴 Backend is DOWN - Starting now...")
        
        # Try to start
        if start_backend():
            logger.info("⏳ Waiting 5 seconds for backend to initialize...")
            time.sleep(5)
            
            # Health check
            if health_check_backend():
                logger.info("✅ Health check PASSED - Backend is responsive")
            else:
                logger.warning("⚠️ Health check FAILED - Backend started but may not be responsive yet")
    else:
        logger.info("✅ Backend is already running")
        
        # Health check even if running
        if not health_check_backend():
            logger.warning("⚠️ Backend not responding - attempting restart...")
            restart_backend()
    
    logger.info("=" * 60)

def market_open_event():
    """Called at 9:00 AM - Market opens"""
    logger.info("\n" + "=" * 60)
    logger.info("⏰ 9:00 AM - MARKET OPEN EVENT")
    logger.info("🟠 PRE-OPEN PERIOD STARTING (9:00-9:15)")
    logger.info("📊 Values will be frozen, no live feed yet")
    logger.info("=" * 60)
    check_and_start_backend()

def live_trading_event():
    """Called at 9:15 AM - Live trading starts"""
    logger.info("\n" + "=" * 60)
    logger.info("⏰ 9:15 AM - LIVE TRADING STARTED")
    logger.info("🟢 LIVE TRADING PERIOD (9:15-3:30)")
    logger.info("📈 Live market values streaming from WebSocket")
    logger.info("=" * 60)

def market_close_event():
    """Called at 3:30 PM - Market closes"""
    logger.info("\n" + "=" * 60)
    logger.info("⏰ 3:30 PM - MARKET CLOSED")
    logger.info("🔴 MARKET CLOSED FOR THE DAY")
    logger.info("📉 Live feed will stop, values will freeze")
    logger.info("=" * 60)

def health_check_scheduled():
    """Periodic health check during market hours"""
    if is_market_day() and is_within_market_hours():
        backend_running = is_backend_running()
        if not backend_running:
            logger.warning("⚠️ Backend crashed during market hours - Restarting...")
            check_and_start_backend()
        else:
            logger.debug("✅ Periodic health check PASSED")

# ============= SCHEDULER SETUP =============

def setup_schedules():
    """Setup all cron jobs"""
    
    # Job 1: Every minute - Check if should start backend
    schedule.every(1).minutes.do(check_and_start_backend)
    logger.info("📅 Scheduled: Check every minute if should start backend")
    
    # Job 2: 9:00 AM on market days - Market open
    schedule.every().monday.at("09:00").do(market_open_event)
    schedule.every().tuesday.at("09:00").do(market_open_event)
    schedule.every().wednesday.at("09:00").do(market_open_event)
    schedule.every().thursday.at("09:00").do(market_open_event)
    schedule.every().friday.at("09:00").do(market_open_event)
    logger.info("📅 Scheduled: Market open at 9:00 AM (Mon-Fri)")
    
    # Job 3: 9:15 AM on market days - Live trading starts
    schedule.every().monday.at("09:15").do(live_trading_event)
    schedule.every().tuesday.at("09:15").do(live_trading_event)
    schedule.every().wednesday.at("09:15").do(live_trading_event)
    schedule.every().thursday.at("09:15").do(live_trading_event)
    schedule.every().friday.at("09:15").do(live_trading_event)
    logger.info("📅 Scheduled: Live trading at 9:15 AM (Mon-Fri)")
    
    # Job 4: 3:30 PM on market days - Market close
    schedule.every().monday.at("15:30").do(market_close_event)
    schedule.every().tuesday.at("15:30").do(market_close_event)
    schedule.every().wednesday.at("15:30").do(market_close_event)
    schedule.every().thursday.at("15:30").do(market_close_event)
    schedule.every().friday.at("15:30").do(market_close_event)
    logger.info("📅 Scheduled: Market close at 3:30 PM (Mon-Fri)")
    
    # Job 5: Health check every 10 minutes during market hours
    schedule.every(10).minutes.do(health_check_scheduled)
    logger.info("📅 Scheduled: Health check every 10 minutes")

def main():
    """Main scheduler loop"""
    logger.info("\n" + "=" * 60)
    logger.info("🚀 MARKET AUTO-START SYSTEM (Python) INITIALIZED")
    logger.info("=" * 60)
    logger.info(f"Project Path: {PROJECT_PATH}")
    logger.info(f"Market Open: {MARKET_OPEN_TIME} AM IST")
    logger.info(f"Pre-Open Period: {MARKET_OPEN_TIME} - {PREOPEN_END_TIME} (Values Frozen)")
    logger.info(f"Live Trading: {LIVE_START_TIME} - {MARKET_CLOSE_TIME} (Live Feed)")
    logger.info(f"Market Days: Monday - Friday (excluding holidays)")
    logger.info(f"Backend Name: {BACKEND_NAME}")
    logger.info(f"Log File: {LOG_FILE}")
    logger.info("=" * 60)
    
    # Setup schedules
    setup_schedules()
    
    # Initial check
    check_and_start_backend()
    
    logger.info("✅ All schedules configured successfully")
    logger.info("💚 System is monitoring for market open time...\n")
    
    # Main loop
    try:
        while True:
            schedule.run_pending()
            time.sleep(1)
    except KeyboardInterrupt:
        logger.info("👋 Scheduler shutting down gracefully...")
        sys.exit(0)

if __name__ == "__main__":
    main()

File: market-auto-start.js
/**
 * CENTRALIZED MARKET AUTO-START SYSTEM
 * ====================================
 * Automatically starts backend at 9 AM on trading days
 * Monitors health and restarts if crashed
 * 
 * Usage on DigitalOcean:
 * 1. pm2 start market-auto-start.js --name "market-scheduler"
 * 2. pm2 save
 * 3. pm2 startup
 * 
 * This will:
 * - Start automatically on server reboot
 * - Monitor backend health 24/7
 * - Auto-restart backend if it crashes
 * - Log all actions for debugging
 */

const cron = require('node-cron');
const { execSync } = require('child_process');
const fs = require('fs');
const path = require('path');

// ============= CONFIGURATION =============
const CONFIG = {
  PROJECT_PATH: '/var/www/mytradingSignal',
  BACKEND_NAME: 'backend',
  FRONTEND_NAME: 'frontend',
  
  // Trading hours (IST - India Standard Time)
  MARKET_OPEN_HOUR: 9,
  MARKET_OPEN_MINUTE: 0,
  MARKET_CLOSE_HOUR: 15,
  MARKET_CLOSE_MINUTE: 30,
  
  // Pre-open period: 9:00 - 9:15 (values frozen)
  PREOPEN_START: 9,
  PREOPEN_END_MINUTE: 15,
  
  // Live values start: 9:15 AM
  LIVE_START_HOUR: 9,
  LIVE_START_MINUTE: 15,
  
  LOG_FILE: '/var/log/mytradingSignal/market-scheduler.log',
  ERROR_LOG: '/var/log/mytradingSignal/market-scheduler-error.log',
};

// ============= MARKET HOLIDAYS (NSE/BSE) =============
// https://www.nseindia.com/market-data/holiday-calendar
const MARKET_HOLIDAYS_2026 = [
  '2026-01-26', // Republic Day
  '2026-03-25', // Holi
  '2026-04-02', // Good Friday
  '2026-05-01', // Labour Day
  '2026-07-17', // Eid-ul-Adha
  '2026-08-15', // Independence Day
  '2026-09-02', // Janmastami
  '2026-10-02', // Gandhi Jayanti
  '2026-10-25', // Dussehra
  '2026-11-01', // Diwali (Sunday)
  '2026-11-11', // Diwali Holidays
  '2026-12-25', // Christmas
];

// Update these before market opens for next year or as needed
const MARKET_HOLIDAYS_2027 = [
  '2027-01-26', // Republic Day
  '2027-03-14', // Holi
  '2027-04-02', // Good Friday
  '2027-05-01', // Labour Day
  '2027-07-06', // Eid-ul-Adha
  '2027-08-15', // Independence Day
  '2027-08-19', // Janmastami
  '2027-10-02', // Gandhi Jayanti
  '2027-10-16', // Dussehra
  '2027-10-20', // Diwali
  '2027-10-21', // Diwali Holidays
  '2027-12-25', // Christmas
];

// ============= LOGGER =============
class Logger {
  constructor(logFile, errorFile) {
    this.logFile = logFile;
    this.errorFile = errorFile;
    this.ensureLogDir();
  }

  ensureLogDir() {
    const dir = path.dirname(this.logFile);
    if (!fs.existsSync(dir)) {
      try {
        fs.mkdirSync(dir, { recursive: true });
      } catch (e) {
        console.error('Failed to create log directory:', e);
      }
    }
  }

  log(message) {
    const timestamp = new Date().toISOString();
    const logMessage = `[${timestamp}] ${message}\n`;
    console.log(logMessage);
    try {
      fs.appendFileSync(this.logFile, logMessage);
    } catch (e) {
      console.error('Failed to write log:', e);
    }
  }

  error(message, err = null) {
    const timestamp = new Date().toISOString();
    const errorMessage = `[${timestamp}] ERROR: ${message}${err ? '\n' + err.toString() : ''}\n`;
    console.error(errorMessage);
    try {
      fs.appendFileSync(this.errorFile, errorMessage);
    } catch (e) {
      console.error('Failed to write error log:', e);
    }
  }
}

const logger = new Logger(CONFIG.LOG_FILE, CONFIG.ERROR_LOG);

// ============= HELPER FUNCTIONS =============

/**
 * Check if today is a trading day
 * Trading days: Monday-Friday, excluding holidays and weekends
 */
function isMarketOpen() {
  const now = new Date();
  const dayOfWeek = now.getDay(); // 0 = Sunday, 5 = Friday, 6 = Saturday
  
  // Skip weekends
  if (dayOfWeek === 0 || dayOfWeek === 6) {
    return false;
  }

  // Check holidays
  const dateStr = now.toISOString().split('T')[0]; // YYYY-MM-DD
  const allHolidays = [...MARKET_HOLIDAYS_2026, ...MARKET_HOLIDAYS_2027];
  
  if (allHolidays.includes(dateStr)) {
    logger.log(`📅 Today is a market holiday: ${dateStr}`);
    return false;
  }

  return true;
}

/**
 * Get current time in IST (UTC+5:30)
 */
function getCurrentTimeIST() {
  const now = new Date();
  // Convert to IST (UTC+5:30)
  const istTime = new Date(now.getTime() + (5.5 * 60 * 60 * 1000));
  return istTime;
}

/**
 * Check if current time is within market hours
 */
function isWithinMarketHours() {
  const istTime = getCurrentTimeIST();
  const hour = istTime.getHours();
  const minute = istTime.getMinutes();
  
  const currentTimeInMinutes = hour * 60 + minute;
  const marketOpenTime = CONFIG.MARKET_OPEN_HOUR * 60 + CONFIG.MARKET_OPEN_MINUTE;
  const marketCloseTime = CONFIG.MARKET_CLOSE_HOUR * 60 + CONFIG.MARKET_CLOSE_MINUTE;
  
  return currentTimeInMinutes >= marketOpenTime && currentTimeInMinutes <= marketCloseTime;
}

/**
 * Check if we're in pre-open period (9:00 - 9:15)
 */
function isPreOpenPeriod() {
  const istTime = getCurrentTimeIST();
  const hour = istTime.getHours();
  const minute = istTime.getMinutes();
  
  return hour === CONFIG.PREOPEN_START && minute < CONFIG.PREOPEN_END_MINUTE;
}

/**
 * Check if backend is running via PM2
 */
function isBackendRunning() {
  try {
    const output = execSync('pm2 list --silent', { encoding: 'utf-8' });
    return output.includes(CONFIG.BACKEND_NAME) && output.includes('online');
  } catch (e) {
    logger.error('Failed to check PM2 list', e);
    return false;
  }
}

/**
 * Start backend using PM2
 */
function startBackend() {
  try {
    logger.log('🚀 Starting backend...');
    // Change to project directory and start backend
    execSync(`cd ${CONFIG.PROJECT_PATH} && pm2 start backend/main.py --name ${CONFIG.BACKEND_NAME} --interpreter python3`, {
      stdio: 'inherit'
    });
    logger.log('✅ Backend started successfully');
    return true;
  } catch (e) {
    logger.error('Failed to start backend', e);
    return false;
  }
}

/**
 * Restart backend using PM2
 */
function restartBackend() {
  try {
    logger.log('🔄 Restarting backend...');
    execSync(`pm2 restart ${CONFIG.BACKEND_NAME}`, { stdio: 'inherit' });
    logger.log('✅ Backend restarted successfully');
    return true;
  } catch (e) {
    logger.error('Failed to restart backend', e);
    return false;
  }
}

/**
 * Health check - HTTP ping to backend
 */
async function healthCheckBackend() {
  try {
    const http = require('http');
    return new Promise((resolve) => {
      const req = http.get('http://localhost:8000/health', (res) => {
        resolve(res.statusCode === 200);
      });
      req.on('error', () => resolve(false));
      req.setTimeout(3000, () => {
        req.destroy();
        resolve(false);
      });
    });
  } catch (e) {
    return false;
  }
}

// ============= MAIN SCHEDULER LOGIC =============

async function checkAndStartBackend() {
  const istTime = getCurrentTimeIST();
  const timeStr = istTime.toLocaleString('en-IN', { timeZone: 'Asia/Kolkata' });
  
  console.log(`\n🕐 Market Scheduler Check - ${timeStr}`);
  
  // Check if it's a market day
  if (!isMarketOpen()) {
    console.log('📊 Market closed (weekend or holiday)');
    return;
  }

  // Check if within market hours
  if (!isWithinMarketHours()) {
    console.log('⏰ Outside market hours');
    return;
  }

  logger.log(`=== MARKET DAY & TRADING HOURS ===`);
  logger.log(`🕐 Current Time (IST): ${timeStr}`);

  // Check if in pre-open period
  if (isPreOpenPeriod()) {
    logger.log('📻 PRE-OPEN PERIOD (9:00-9:15): Values frozen, no live feed');
  } else {
    logger.log(`📈 LIVE TRADING PERIOD (9:15-3:30): Live values streaming`);
  }

  // Check if backend is running
  const backendRunning = isBackendRunning();
  
  if (!backendRunning) {
    logger.log('🔴 Backend is DOWN - Starting now...');
    
    // Try to start backend
    const started = startBackend();
    
    if (started) {
      logger.log('⏳ Waiting 5 seconds for backend to initialize...');
      await new Promise(r => setTimeout(r, 5000));
      
      // Health check
      const healthy = await healthCheckBackend();
      if (healthy) {
        logger.log('✅ Health check PASSED - Backend is responsive');
      } else {
        logger.log('⚠️  Health check FAILED - Backend started but not responsive yet');
      }
    }
  } else {
    logger.log('✅ Backend is already running');
    
    // Periodic health check even if running
    const healthy = await healthCheckBackend();
    if (!healthy) {
      logger.log('⚠️  Backend not responding - attempting restart...');
      restartBackend();
    }
  }
}

// ============= CRON JOBS =============

function setupSchedules() {
  // Job 1: Check every minute if we should start backend
  cron.schedule('* * * * *', checkAndStartBackend, {
    timezone: 'Asia/Kolkata' // IST timezone
  });
  logger.log('📅 Cron Job 1: Minute-level monitoring (every 60 seconds)');

  // Job 2: At exactly 9:00 AM IST - Force start backend
  cron.schedule('0 9 * * 1-5', async () => {
    logger.log('\n⏰ 9:00 AM Market Open Event Triggered');
    logger.log('🟠 PRE-OPEN PERIOD STARTING (9:00-9:15)');
    logger.log('📊 Values will be frozen, no live feed yet');
    logger.log('Attempting to start backend for pre-open...');
    await checkAndStartBackend();
  }, {
    timezone: 'Asia/Kolkata'
  });
  logger.log('📅 Cron Job 2: 9:00 AM Daily Trigger (Mon-Fri)');

  // Job 3: At 9:15 AM IST - Live trading starts
  cron.schedule('15 9 * * 1-5', () => {
    logger.log('\n⏰ 9:15 AM - Live Trading Started');
    logger.log('🟢 LIVE TRADING PERIOD (9:15-3:30)');
    logger.log('📈 Live market values streaming from WebSocket');
  }, {
    timezone: 'Asia/Kolkata'
  });
  logger.log('📅 Cron Job 3: 9:15 AM Daily Alert (Mon-Fri)');

  // Job 4: At 3:30 PM IST - Market close
  cron.schedule('30 15 * * 1-5', () => {
    logger.log('\n⏰ 3:30 PM - Market Closed');
    logger.log('🔴 MARKET CLOSED');
    logger.log('📉 Live feed will stop, values will freeze');
  }, {
    timezone: 'Asia/Kolkata'
  });
  logger.log('📅 Cron Job 4: 3:30 PM Daily Alert (Mon-Fri)');

  // Job 5: Health check every 10 minutes during market hours
  cron.schedule('*/10 9-15 * * 1-5', async () => {
    if (isWithinMarketHours()) {
      const backendRunning = isBackendRunning();
      if (!backendRunning) {
        logger.log('⚠️  Backend crashed during market hours - Restarting...');
        await checkAndStartBackend();
      }
    }
  }, {
    timezone: 'Asia/Kolkata'
  });
  logger.log('📅 Cron Job 5: Health Check Every 10 Minutes (Market Hours)');
}

// ============= STARTUP =============

logger.log('\n' + '='.repeat(60));
logger.log('🚀 MARKET AUTO-START SYSTEM INITIALIZED');
logger.log('='.repeat(60));
logger.log(`Project Path: ${CONFIG.PROJECT_PATH}`);
logger.log(`Market Open: 9:00 AM IST`);
logger.log(`Pre-Open Period: 9:00 - 9:15 (Values Frozen)`);
logger.log(`Live Trading: 9:15 AM - 3:30 PM (Live Feed)`);
logger.log(`Market Days: Monday - Friday (excluding holidays)`);
logger.log(`Backend Name: ${CONFIG.BACKEND_NAME}`);
logger.log(`Log File: ${CONFIG.LOG_FILE}`);
logger.log('='.repeat(60) + '\n');

// Setup all cron jobs
setupSchedules();

// Initial check
checkAndStartBackend();

// Graceful shutdown
process.on('SIGINT', () => {
  logger.log('\n👋 Scheduler shutting down gracefully...');
  process.exit(0);
});

process.on('SIGTERM', () => {
  logger.log('\n👋 Scheduler received termination signal...');
  process.exit(0);
});

logger.log('✅ All cron jobs scheduled successfully');
logger.log('💚 System is monitoring for market open time...');
